src/site/content/en/blog/form-fatigue-validate-user-input/index.md

```diff
@@ -0,0 +1,308 @@
+{% endAside %}
+
+## Use the `pattern` attribute
+
+In addition to restricting input, you can validate input against a regular
+expression using the `pattern` attribute. For example, a valid US zip code is 5
+digits and sometimes followed by a dash and an additional 4 digits:
+
+
+```html  
+<input type="text" pattern="^\d{5,6}(?:[-\s]\d{4})?$" ...>  
+```
+
+The following regular expression validates credit card numbers:
+
+
+```html
+^(?:4[0-9]{12}(?:[0-9]{3})?|5[1-5][0-9]{14}|3[47][0-9]{13}|3(?:0[0-5]|[68][0-9])[0-9]{11}|6(?:011|5[0-9]{2})[0-9]{12}|(?:2131|1800|35\d{3})\d{11})$  
+```
+
+{% Aside 'note' %}
+Unlike `pattern`, `min`, `max`, `minlength` and `maxlength` will restrict user input in real-time.
+{% endAside %}
+
+## Custom validation
+
+If built-in validation attributes aren't enough, use the
+[Constraint Validation API](https://w3c.github.io/html/sec-forms.html#constraints)
+to manually mark fields invalid. To manually invalidate a field, use
+`setCustomValidity()` to pass a custom validation message to the invalid input
+element. 
+
+For example, you can immediately validate a shipping address using a back-end location service:
+
+```js/3
+let shippingAddress = shippingAddressInput.value;  
+if (!checkValidAddress(shippingAddress)) {  
+  // invalid card number, set custom validity message  
+  shippingAddressInput.setCustomValidity('This is not a valid address.');  
+} else {  
+  // valid card number, reset message  
+  shippingAddressInput.setCustomValidity('');  
+}  
+```  
+Be sure to reset the error message when the field becomes valid:  
+
+<!-- TODO (mdiblasio): how to bold in templates?  -->
+
+```js/6
+let shippingAddress = shippingAddressInput.value;  
+if (!checkValidAddress(shippingAddress)) {  
+  // invalid card number, set custom validity message  
+  shippingAddressInput.setCustomValidity('This is not a valid address.');  
+} else {  
+  // valid card number, reset message  
+  shippingAddressInput.setCustomValidity('');  
+}  
+``` 
+
+### Set custom validation messages
+
+Default validation messages can be vague and difficult to understand. Use the [Constraint Validation API](https://w3c.github.io/html/sec-forms.html#constraints) to provide helpful, specific and non-technical validation messages when users enter invalid information.
+
+For example, you can provide different email validation messages based on a
+particular invalidation reason using the `validity` object property:
+
+```js/2,5
+if (!emailInput.validity.valid) {  
+  // case 1 - value missing
+  if (emailInput.validity.valueMissing)  
+    emailInput.setCustomValidity('We need your email to complete your purchase');  
+  // case 2 - pattern mismatch
+  else if (emailInput.validity.patternMismatch)  
+    emailInput.setCustomValidity('Your email address isn\'t valid (e.g. name@gmail.com)');  
+  // ...   
+} else {  
+  emailInput.setCustomValidity('');  
+}  
+```
+
+#### Case 1: Missing required email address
```

Comment:
Are the case 1 and case 2 examples different enough that you need both?
I was wondering if it might be cool to embed a running Glitch that lets people play with the code above.

Reply:
good called, consolidated. added a TODO to add glitch sample in future